### PR TITLE
be generic about the thing to write the path into

### PR DESCRIPTION
**Pull request by jordens**
The downside is that we don't check capacity anymore.
For most things that are `Write` they don't have a capacity.
On the other hand this was always done redundantly also in the mqtt client.

### CODE REVIEW

File: tests/iter.rs [hunk 1]
```diff
@@ -21,10 +21,10 @@ fn insufficient_space() {
     assert_eq!(meta.count, 3);
 
     // Ensure that we can't iterate if we make a state vector that is too small.
-    assert!(Settings::iter_paths::<1, 256>().is_err());
+    assert!(Settings::iter_paths::<1, String<256>>().is_err());
 
     // Ensure that we can't iterate if the topic buffer is too small.
-    assert!(Settings::iter_paths::<10, 1>().is_err());
+    // assert!(Settings::iter_paths::<10, String<1>>().is_err());
```
Comment (ryan-summers) [inline tests/iter.rs:27]:
```suggestion
    assert!(Settings::iter_paths::<10, String<1>>().is_err());
```

Comment (ryan-summers) [reply tests/iter.rs:27]: Or is the problem that we no longer detect this as an error? That seems like a problem

File: src/lib.rs [hunk 2]
```diff
@@ -198,12 +198,11 @@ pub trait SerDe<S>: Miniconf {
     ///
     /// # Generics
     /// * `L`  - The maximum depth of the path, i.e. number of separators plus 1.
-    /// * `TS` - The maximum length of the path in bytes.
+    /// * `P`  - The type to hold the path.
     ///
     /// # Returns
     /// A [MiniconfIter] of paths or an [IterError] if `L` or `TS` are insufficient.
```
Comment (ryan-summers) [inline src/lib.rs:204]:
```suggestion
    /// A [MiniconfIter] of paths or an [IterError] if `P` or `TS` are insufficient.
```

File: src/iter.rs [hunk 3]
```diff
@@ -24,26 +22,19 @@ pub struct MiniconfIter<M: ?Sized, const L: usize, const TS: usize, S> {
...
     }
 }
 
-impl<M: ?Sized + Miniconf + SerDe<S>, const L: usize, const TS: usize, S>
-    MiniconfIter<M, L, TS, S>
-{
+impl<M: SerDe<S>, S, const L: usize, P> MiniconfIter<M, S, L, P> {
     pub fn metadata() -> Result<Metadata, IterError> {
         let meta = M::metadata(M::SEPARATOR.len_utf8());
-        if TS < meta.max_length {
```
Comment (ryan-summers) [inline src/iter.rs:43]: If we can't pre-emptively check the length, do we even want to precalculate the data length? I guess it's still potentially useful to the end user for initial setup.

File: tests/iter.rs (hunk 1 above)
Comment (jordens) [reply tests/iter.rs:27]: See the PR description.
I think it's ok to skip it as we've always found it necessary to do a similar check and redundant check elsewhere as well.

Comment (ryan-summers) [reply tests/iter.rs:27]: Oh sorry, I thought this was related to the first iteration (next()) not returning an error. The construction not returning an error is fine. Can we remove the broken test?

**Review by ryan-summers — APPROVED** [review]

**Pull request merged**